addons/print/models/print_model.py

```diff
@@ -0,0 +1,295 @@
+                })
+            else: # error : not attachament can be generate, no attach_id or no res_model/res_id
+                current_order.write({
+                    'state' : 'error',
+                    'error_message' : _('The document has no associated PDF : you have to give select an Attachment file, or set up the Object ID and Model Name fields.')
+                })
+
+    def _prepare_printing(self):
+        """ Prepare the orders for delivery. It executes the operations to put
+            them into the 'ready_to_send' state (or 'error' if something wrong happens).
+            To allow optimizations in the provider implementation, the orders
+            are grouped by provider.
+        """
+        # generate PDF for the recordset
+        self._generate_attachment()
+
+        providers = set(self.mapped('provider_id.id'))
+        for provider_id in providers: # process by provider id
+            records = self.filtered(lambda r: r.provider_id.id == provider_id)
+            # call provider implementation
+            provider_name = records[0].provider_id.provider
+            if hasattr(records, '_%s_prepare_printing' % provider_name):
+                getattr(records, '_%s_prepare_printing' % provider_name)()
+
+    def _deliver_printing(self):
+        """ Send the orders for delivery to the Provider. It executes the operations to put
+            them into the 'sent' state (or 'error' if something wrong happens).
+            Required : the print.order must be in the 'ready_to_send' state.
+            This method group print.order by provider (type), to stay the optimized.
+        """
+        providers = set(self.mapped('provider_id.provider'))
+        for provider_name in providers: # process by provider type
+            if hasattr(self, '_%s_deliver_printing' % provider_name):
+                records = self.filtered(lambda r: r.provider_id.provider == provider_name)
+                getattr(records, '_%s_deliver_printing' % provider_name)()
+
+    def _validate_printing(self):
+        """ For the given recordset, apply the action on the printable objects when the sending
+            is correctly done. This call the 'print_validate_sending' method of the printable
+            object (see print.mixin below).
+            This method group the PO by res_model to optimize the browse.
+        """
+        # group the PO by res_model (res_model is optional field, so only keep PO where res_model not None/False)
+        res_models = set(self.filtered(lambda r: r.res_model).mapped('res_model'))
+        for model in res_models:
+            if hasattr(self.env[model], 'print_validate_sending'):
+                objects = self.env[model].browse(self.filtered(lambda r: r.res_model == model).mapped('res_id'))
+                objects.print_validate_sending()
+
+    @api.multi
+    def process_order_queue(self, order_ids=None):
```

Comment:
Is it truly a multi ? As order_ids is given, and if void it does not use self.ids but search for not sent orders. It seems more like a model method to me.

BTW, is the decorator necessary, as this is a new module and new code ?